lib/components/user/notification-prefs-pane.js

```diff
@@ -41,16 +76,119 @@ class NotificationPrefsPane extends Component {
+    this.setState({
+      isPhoneFieldModified: false,
+      isVerificationPending: false
+    })
+
+    this.props.onRevertUserPhoneNumber()
+  }
+
+  _handleRequestPhoneVerificationCode = () => {
+    // Send phone verification request.
+    this.props.onRequestPhoneVerificationCode()
+
+    // Prompt for code.
+    this.setState({
+      isVerificationPending: true,
+      phoneValidationCode: ''
+    })
+  }
+
+  _handleSendPhoneVerificationCode = async () => {
+    await this.props.onSendPhoneVerificationCode(this.state.phoneValidationCode)
+
+    // Exit verification modal and update "initialPhoneNumber" if isPhoneNumberVerified was changed to true.
+    // Erase the verification code in all cases.
+    if (this.props.userData.isPhoneNumberVerified) {
+      this.setState({
+        initialPhoneNumber: this.props.userData.phoneNumber,
+        isVerificationPending: false,
+        phoneValidationCode: ''
+      })
+    } else {
+      this.setState({
+        phoneValidationCode: ''
+      })
+    }
   }
 
   render () {
     const { userData } = this.props
+    const {
+      initialPhoneNumber,
+      isPhoneFieldModified,
+      isVerificationPending,
+      phoneValidationCode
+    } = this.state
     const {
       email,
+      isPhoneNumberVerified,
       notificationChannel,
       phoneNumber
     } = userData
 
+    // Here are the states we are dealing with:
+    // - First time entering a phone number (phoneNumber blank, not modified)
+    //   => no color, no feedback indication.
+    // - Typing backspace all the way to erase a number (phoneNumber blank, modified)
+    //   => red, "[X] Please provide a number" indication.
+    // - Viewing a verified phone number (phoneNumber non-blank, same as initial, verified)
+    //   => green, "[V] Verified" indication.
+    // - Editing a phone number (phoneNumber non-blank, different than initial or not verified)
+    //   => yellow, "[!] Verification required" indication.
+
+    const isPhoneNumberBlank = !(phoneNumber && phoneNumber.length)
+    const isPhoneNumberSameAsInitial = phoneNumber === initialPhoneNumber
+
+    let phoneStatusGlyph // one of the Bootstrap glyphs.
+    let phoneStatusText
+    let phoneFieldValidationState // one of the Bootstrap validationState values.
+    let shouldVerifyPhone = false
+
+    if (isPhoneNumberBlank) {
+      if (!isPhoneFieldModified) {
+        // Do not show an indication.
+      } else {
+        phoneStatusGlyph = 'remove'
+        phoneStatusText = 'Please provide a number'
+        phoneFieldValidationState = 'error'
+      }
+    } else if (isPhoneNumberSameAsInitial && isPhoneNumberVerified) {
+      phoneStatusGlyph = 'ok'
+      phoneStatusText = 'Verified'
+      phoneFieldValidationState = 'success'
+    } else {
+      phoneStatusGlyph = 'warning-sign'
+      phoneStatusText = 'Verification required'
+      phoneFieldValidationState = 'warning'
+      shouldVerifyPhone = true
+    }
```

Comment:
These states aren't working perfectly for me. E.g., I can click "Revert" twice in a row for some reason?

Reply:
Removed